src/hotspot/share/oops/instanceKlass.cpp

```diff
@@ -2665,12 +2665,14 @@ static void method_release_C_heap_structures(Method* m) {
 }
 
 // Called also by InstanceKlass::deallocate_contents, with false for release_constant_pool.
-void InstanceKlass::release_C_heap_structures(bool release_constant_pool) {
+void InstanceKlass::release_C_heap_structures(bool release_sub_metadata) {
   // Clean up C heap
   Klass::release_C_heap_structures();
 
   // Deallocate and call destructors for MDO mutexes
-  methods_do(method_release_C_heap_structures);
+  if (release_sub_metadata) {
+    methods_do(method_release_C_heap_structures);
+  }
```

Comment:
I don't understand this aspect of the change. It seems unrelated to the missing destructor issue and seems to cause a significant change in behaviour as we will no longer call `methods_do` when called from `deallocate_contents` with the renamed `false` argument.

Reply:
This boolean prevents us from calling the MethodData destructor twice and releasing MethodData C heap memory twice.  InstanceKlass::deallocate_contents calls release_C_heap_structures with it false, and then calls through the Metadata pointer hierarchy to MetadataFactory::free_metadata, which calls the MethodData destructor and MethodData::deallocate_contents which calls MethodData::release_C_heap_structures().   This is the delegation scheme for releasing metadata between InstanceKlass and ConstantPool.

InstanceKlass::release_C_heap_structures() is also called during class unloading and must walk the whole metadata, releasing C heap memory.  It's passed true. It doesn't go through deallocate_contents because the metadata is freed together at the same time with the metaspace that contains it.

Reply:
Thanks Coleen I see now how the code has been shuffled around:

Old code:
```
InstanceKlass::deallocate_contents
-> release_C_heap_structures(/* release_sub_metadata */ false);
  -> methods_do(method_release_C_heap_structures);
    -> m->release_C_heap_structures
      -> method_data()->~MethodData();
-> deallocate_methods(loader_data, methods());
  -> MetadataFactory::free_metadata(loader_data, method);
```
New code:
```
InstanceKlass::deallocate_contents
-> release_C_heap_structures(/* release_sub_metadata */ false);
-> deallocate_methods(loader_data, methods());
  -> MetadataFactory::free_metadata(loader_data, method);
    -> method_data()->~MethodData();
```

Reply:
Yes, plus releasing the JVMCI thing that MethodData points to also.